lightningd/invoice.c

```diff
@@ -356,33 +388,51 @@ static struct route_info **select_inchan(const tal_t *ctx,
 			continue;
 		}
 
-		/* Even after reserve taken into account */
-		if (!amount_msat_sub_sat(&avail,
-					 avail, c->our_config.channel_reserve))
+		/* Even after taken into account their reserve */
+		if (!amount_msat_sub_sat(&capacity_to_pay_us, their_msat,
+				c->our_config.channel_reserve))
 			continue;
 
-		if (!amount_msat_sub(&excess, avail, capacity_needed))
+		if (!amount_msat_sub(&excess, capacity_to_pay_us, amount_needed))
 			continue;
 
+		/* Excess as percentage of capacity, assumes check_config_bounds */
+		excess_perc = (double)excess.millisatoshis * 0.1 /               /* Raw:excess_perc */
+			  (c->funding.satoshis                                       /* Raw:excess_perc */
+			  - c->our_config.channel_reserve.satoshis                   /* Raw:excess_perc */
+			  - c->channel_info.their_config.channel_reserve.satoshis);  /* Raw:excess_perc */
+
 		/* Is it offline? */
 		if (c->owner == NULL) {
 			*any_offline = true;
 			continue;
 		}
 
-		/* Avoid divide-by-zero corner case. */
-		wsum += excess.millisatoshis + 1; /* Raw: rand select */
-		if (pseudorand(1ULL << 32)
-		    <= ((excess.millisatoshis + 1) << 32) / wsum) /* Raw: rand select */
-			r = &inchans[i];
+		sample.route = &inchans[i];
+		sample.weight = excess_perc;
+		tal_arr_expand(&S, sample);
 	}
 
-	if (!r)
+	if (!tal_count(S))
 		return NULL;
 
-	ret = tal_arr(ctx, struct route_info *, 1);
-	ret[0] = tal_dup(ret, struct route_info, r);
-	return ret;
+	/* Use weighted reservoir sampling, see:
+	 * https://en.wikipedia.org/wiki/Reservoir_sampling#Algorithm_A-Chao
+	 * But (currently) the result will consist of only one sample (k=1) */
+	R = tal_arr(ctx, struct route_info *, 1);
+	R[0] = tal_dup(R, struct route_info, S[0].route);
+	wsum = S[0].weight;
+
+	for (size_t i = 1; i < tal_count(S); i++) {
+		wsum += S[i].weight;
+		p = S[i].weight / wsum;
+		double random_1 = (double)pseudorand(UINT64_MAX) / UINT64_MAX;
```

Comment:
We should expose (separate commit I'm afraid!) pseudorand_double() which calls isaac64_next_double() underneath.  Much nicer than this...